### PR TITLE
Add project search index with permission mirroring

### PR DESCRIPTION
**Pull request by kbirk**

### CODE REVIEW

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ProjectController.java [hunk 1]
```diff
@@ -1120,6 +1130,9 @@ public ResponseEntity<JsonNode> updateProjectUserPermissions(
 
 			final RebacProject what = new RebacProject(projectId, reBACService);
 			final RebacUser who = new RebacUser(userId, reBACService);
+
+			// no need to update the search perms since we cant never reduce below a read.
```
Comment (dgauldie) [inline packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ProjectController.java:1134]: double negative

File: packages/server/src/main/resources/static/es/index-templates/tds_2.0_project_index_template.json [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            },
+            "vector": {
+              "type": "dense_vector",
+              "element_type": "float",
+              "dims": 1536
+            },
+            "embeddingId": {
+              "type": "keyword"
+            },
+            "span": {
```
Comment (mwdchang) [inline packages/server/src/main/resources/static/es/index-templates/tds_2.0_project_index_template.json:61]: What are `embeddingId` and `span`?

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ProjectController.java (hunk 1 above)
Comment (kbirk) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/controller/dataservice/ProjectController.java:1134]: I've never not heard something that doesn't say that this isn't what you don't mean, or isn't it?

File: packages/server/src/main/resources/static/es/index-templates/tds_2.0_project_index_template.json (hunk 2 above)
Comment (kbirk) [reply packages/server/src/main/resources/static/es/index-templates/tds_2.0_project_index_template.json:61]: The embedding generation can only take an input of a fixed size. So for large inputs we chunk them. Then the span a two element array containing the start / stop index in the source that the embedding references. 

So the id + span allows us the option to known exactly which chunk of text the lookup matches.

**Review by YohannParis — APPROVED** [review]

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+		elasticService.deleteIndex(index);
+	}
+
+	public void indexProject(final Project project) throws IOException {
+		final ProjectDocument doc = new ProjectDocument();
+		doc.setUserId(project.getUserId());
+		doc.setPublicAsset(project.getPublicAsset());
+		doc.setCreatedOn(new Timestamp(System.currentTimeMillis()));
+		doc.setUpdatedOn(new Timestamp(System.currentTimeMillis()));
+		doc.setPermissionJoin(PermissionJoin.CreateProjectJoin());
```
Comment (bigglesandginger) [inline packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java:125]: So this is the initial "permission" for a project, but it doesn't have a user assigned?  Why isn't the creator added to the list?

**Comment by bigglesandginger** [comment]
The use of the word `permission` I think should be renamed to `canRead` or something like that as it does not track the Permission enum and instead is simply a list of users that can know of the project

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		return elasticService.getIndexFromAlias(getAlias());
+	}
+
+	@Data
+	public static class PermissionJoin {
+
+		private String name = null;
+		private UUID parent = null;
+
+		static PermissionJoin CreateProjectJoin() {
```
Comment (bigglesandginger) [inline packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java:50]: should this not be `createProjectJoin()` as it is not the constructor for PermissionJoin

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+			final PermissionJoin join = new PermissionJoin();
+			join.setName("project");
+			join.setParent(null);
+			return join;
+		}
+
+		static PermissionJoin CreatePermissionJoin(final UUID projectId) {
```
Comment (bigglesandginger) [inline packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java:57]: should this not be `createPermissionJoin()` as it is not the constructor for PermissionJoin

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+		elasticService.indexWithRouting(getAlias(), project.getId().toString(), doc, routing);
+	}
+
+	public void updateProject(final Project project) throws IOException {
+		final ProjectDocument doc = new ProjectDocument();
+		doc.setUserId(project.getUserId());
+		doc.setPublicAsset(project.getPublicAsset());
+		doc.setUpdatedOn(new Timestamp(System.currentTimeMillis()));
+		doc.setPermissionJoin(PermissionJoin.CreateProjectJoin());
+
```
Comment (bigglesandginger) [inline packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java:138]: Will this not erase the existing `PermissionJoin` thing?

Comment (bigglesandginger) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java:138]: Or, I don't understand this data structure.  It appears that there is one PermissionJoin for a project and yet it changes to be either a specific user via `addProjectPermission` or reset to whatever `CreateProjectJoin()` creates...

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java (hunk 3 above)
Comment (kbirk) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java:125]: That is the permission "join field" declaring that the document is a `"project"` document, and a parent in the parent-child relationship.

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java (hunk 6 above)
Comment (kbirk) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java:138]: The two types of documents share the same index. So the "parent", which is the project, has a join field declaring it a project parent type. And the "child", which is the user read permission, has a join field declaring it a child doc. The `userId` field is overrided, in projects it is the owner of the project, and in a permission it is the userId of the person who has read permission.

Comment (kbirk) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java:138]: The structure of join fields is not very intuitive. They aren't separate docs in different indices, they are in the same index, and have the same mapping. The implicit `name` and `parent` property of the join field are used to define the relationship

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/elasticsearch/ElasticsearchService.java
[the remark this answers is not in the document]
Comment (kbirk) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/elasticsearch/ElasticsearchService.java:1]: My editorconfig is the default that is in the project :(

[the remark this answers is not in the document]
Comment (YohannParis) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/elasticsearch/ElasticsearchService.java:1]: _C'est la vie !_

File: packages/server/src/main/resources/static/es/index-templates/tds_2.0_project_index_template.json (hunk 2 above)
Comment (mwdchang) [reply packages/server/src/main/resources/static/es/index-templates/tds_2.0_project_index_template.json:61]: gotcha, thanks!

File: packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java (hunk 6 above)
Comment (bigglesandginger) [reply packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/ProjectSearchService.java:138]: Well based on the created_on fiasco... I suspect this is where this code will fail

**Review by bigglesandginger — APPROVED** [review]
As we do not have groups, this should work fine (assuming PermissionJoin works as advertised).  If we ever add groups/sub-groups then this will need a good sanding, but the concept as I take it should be solid.

**Pull request merged**